macaroonbakery/store.py

```diff
@@ -25,21 +26,52 @@ def get_ops(self, key):
         return ops
 
 
-class MemoryKeyStore:
+class RootKeyStore(object):
+    ''' Defines store for macaroon root keys.
+    '''
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def get(self, id):
+        ''' Returns the root key for the given id.
+        If the item is not there, it returns None.
+        @param id: bytes
+        @return: bytes
+        '''
+        raise NotImplementedError('get method must be defined in '
+                                  'subclass')
+
+    @abc.abstractmethod
+    def root_key(self):
+        ''' Returns the root key to be used for making a new macaroon, and an
+        id that can be used to look it up later with the get method.
+        Note that the root keys should remain available for as long as the
+        macaroons using them are valid.
+        Note that there is no need for it to return a new root key for every
+        call - keys may be reused, although some key cycling is over time is
+        advisable.
+        @return: bytes
+        '''
+
+
+class MemoryKeyStore(RootKeyStore):
     ''' MemoryKeyStore returns an implementation of
     Store that generates a single key and always
     returns that from root_key. The same id ("0") is always
     used.
     '''
-    def __init__(self):
-        self.key = None
+    def __init__(self, key=None):
+        '''
+        @param key: bytes
+        '''
+        if key is None:
+            key = os.urandom(24)
+        self._key = key
 
     def get(self, id):
-        if len(id) != 1 or id[:1] != b'0' or self.key is None:
+        if len(id) != 1 or id[:1] != b'0' or self._key is None:
```

Comment:
_key should never be None AFAICS.

That is, I think:

    if id != b'0':
        return None

is probably sufficient.